### PR TITLE
feature: start working on FastHTML example

### LINKED ISSUE #14: add lesson using FastHTML
[FastHTML](https://docs.fastht.ml/) pulls together routing and HTMX in interesting ways.

### PR DESCRIPTION
**Pull request by gvwilson**
Lots still to do, especially around error handling, but it's a start.

closes #14

### CODE REVIEW

File: 22_fasthtml/server.py
Comment (juananpe) [inline 22_fasthtml/server.py:1]: I'm not familiar with fastHTML, but having a look to the fastHTML documentation it seems to me that the parameters of the POST /login handler should be treated/passed in a different way. And... is request an object known to fastHTML?

**Review by juananpe — CHANGES_REQUESTED** [review]
The fixes in files not related to FastHTML are OK.

The FastHTML examples should be reviewed (they are not working as expected). 

The login form is missing and the logic to manage the login is still using code from Flask.
I'm still learning about the basis of FastHTML so my fixes might be totally wrong, but here is my attempt:
https://gist.github.com/juananpe/14a8abb5d22fe0f381afda775880c19d

**Comment by gvwilson** [comment]
I moved the minor changes to #16, which ought to be good to merge. And yes, this branch still doesn't have login form, JWT, etc. - I'm learning my way around FastHTML as I go. (This branch is now off the one used for #16, so we should merge that before worrying about diffs.)

**Comment by juananpe** [comment]
> I moved the minor changes to #16, which ought to be good to merge. And yes, this branch still doesn't have login form, JWT, etc. - I'm learning my way around FastHTML as I go. (This branch is now off the one used for #16, so we should merge that before worrying about diffs.)

Do you know if I can apply the sugested changes directly to this PR, or are file edits restricted to the PR owner?

**Comment by gvwilson** [comment]
You should be able to `git fetch origin` and then `git checkout 14-fasthtml` to get the branch in its current state on your machine, then edit and push your changes to the same branch to update this PR.

**Comment by juananpe** [comment]
> You should be able to `git fetch origin` and then `git checkout 14-fasthtml` to get the branch in its current state on your machine, then edit and push your changes to the same branch to update this PR.

Thanks, it was easier than I thought. I've added a new commit to this PR with the new version of server.py and views.py adding authentication support; feel free to remove it entirely, it is just a proof of concept.

**Comment by gvwilson** [comment]
@juananpe thanks for adding the JWT stuff - if you look at the commit comment on 627f99a, I'm displaying the user name when the user is logged in:
    
1.  Get the staff ID from the JWT if it's available.
1.  Pass it to the view.
1.  If the staff ID matches data in one of the records, pull out personal and family names.
1.  Use Alpine to display these conditionally in the page.
    
What it *should* do is get the user IDs in the model query along with the names. Alternatively, skip Alpine entirely and only add the username to the page if the staff ID is present and found in the data.

What do you think? Handle display purely on the server side instead of mixing Alpine into this?

**Comment by juananpe** [comment]
> What it _should_ do is get the user IDs in the model query along with the names. Alternatively, skip Alpine entirely and only add the username to the page if the staff ID is present and found in the data.
> What do you think? Handle display purely on the server side instead of mixing Alpine into this?

I would go to the more straightforward approach of handling the display on the server side, eliminating the need for Alpine.js. You told me to keep things simple :)

In my new commit I've implemented that approach (removing Alpine.js code) and added an option to logout.

**Comment by gvwilson** [comment]
thanks @juananpe - I don't see your new commit in this branch - did you push it to GitHub? Cheers - @gvwilson

**Comment by juananpe** [comment]
> thanks @juananpe - I don't see your new commit in this branch - did you push it to GitHub? Cheers - @gvwilson

Ops! My fault. I pushed to the wrong branch :-\ 
https://github.com/lessonomicon/webonomicon/compare/main...pr-15
Now the commit should be here, in `14-fasthtml`

**Comment by gvwilson** [comment]
thank you!

**Pull request merged**